geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/KryoLazyDensityCoprocessor.scala

```diff
@@ -0,0 +1,168 @@
+    var scanner : InternalScanner = null
+    try {
+      val scan = new Scan
+      val options: Map[String, String] = deserializeOptions(byteFilterArray)
+      val sft = SimpleFeatureTypes.createType("input", options(SFT_OPT))
+      val serializer = new KryoFeatureSerializer(sft, SerializationOptions.withoutId)
+      val densityResult: DensityResult = this.init(options, sft)
+      scanner = env.getRegion.getScanner(scan)
+      val results = new java.util.ArrayList[Cell]
+      var hasMore = false
+      do {
+        hasMore = scanner.next(results)
+        for (cell <- results) {
+          val sf = serializer.deserialize(cell.getValueArray, cell.getValueOffset, cell.getValueLength)
+          aggregateResult(sf, densityResult)
+        }
+        results.clear()
+      } while (hasMore)
+
+      val result: Array[Byte] = KryoLazyDensityUtils.encodeResult(densityResult)
+      response = DensityResponse.newBuilder.setSf(ByteString.copyFrom(result)).build
+    } catch {
+      case ioe: IOException => {
+        ResponseConverter.setControllerException(controller, ioe)
+      }
+      case cnfe: ClassNotFoundException => {
+        logger.error(Throwables.getStackTraceAsString(cnfe))
+      }
+      case dse: DeserializationException => {
+        logger.error(Throwables.getStackTraceAsString(dse))
+      }
+    } finally IOUtils.closeQuietly(scanner)
+    done.run(response)
+  }
+}
+
+object KryoLazyDensityCoprocessor extends KryoLazyDensityUtils{
+
+  final private val logger = LoggerFactory.getLogger(classOf[KryoLazyDensityCoprocessor])
+
+  private val SFT_OPT = "sft"
+
+  /**
+    * Creates an iterator config for the kryo density iterator
+    */
+  def configure(sft: SimpleFeatureType,
+                filter: Option[Filter],
+                hints: Hints): Map[String, String] = {
+    import org.locationtech.geomesa.index.conf.QueryHints.RichHints
+    val envelope = hints.getDensityEnvelope.get
+    val (width, height) = hints.getDensityBounds.get
+    val weight = hints.getDensityWeight
+    configure(sft, filter, envelope, width, height, weight)
+  }
+
+  protected def configure(sft: SimpleFeatureType,
+                                     filter: Option[Filter],
+                                     envelope: Envelope,
+                                     gridWidth: Int,
+                                     gridHeight: Int,
+                                     weightAttribute: Option[String]): Map[String, String] = {
+    val is = mutable.Map.empty[String, String]
+    is.put(ENVELOPE_OPT, s"${envelope.getMinX},${envelope.getMaxX},${envelope.getMinY},${envelope.getMaxY}")
+    is.put(GRID_OPT, s"$gridWidth,$gridHeight")
+    weightAttribute.foreach(is.put(WEIGHT_OPT, _))
+    is.put(SFT_OPT, SimpleFeatureTypes.encodeType(sft, false))
+    is.toMap
+  }
+
+  @throws[IOException]
+  def serializeOptions(map: Map[String, String]): Array[Byte] = {
+    val fis = new ByteArrayOutputStream
```

Comment:
[picky] rename ```fis``` to maybe ```baos``` since ```fis``` usually means file input stream. prob also rename ```ois``` below for good measure (is = input stream)